File: code/cpp/2265.cpp
//https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    struct avg { int sum; int count; };

    avg foo(TreeNode* root) {
        if (!root) return { 0, 0 };
        avg r = foo(root->right),
            l = foo(root->left);
        int _sum = r.sum + l.sum + root->val,
            _count = r.count + l.count + 1;
        if (root->val == _sum / _count)
            res++;
        return {_sum, _count};
    }

    int averageOfSubtree(TreeNode* root) {
        foo(root);
        return res;
    }
};

File: code/python/2265.py
# https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        self.res = 0

        def foo(root):
            if (root == None): return 0, 0
            r_sum, r_cnt = foo(root.right)
            l_sum, l_cnt = foo(root.left)
            _sum = r_sum + l_sum + root.val
            _count = r_cnt + l_cnt + 1
            if root.val == _sum // _count:
                self.res += 1
            return _sum, _count

        foo(root)
        return self.res
